packages/material-ui/src/Tooltip/Tooltip.js

```diff
@@ -72,120 +73,163 @@ export const styles = theme => ({
   },
 });
 
-class Tooltip extends React.Component {
-  ignoreNonTouchEvents = false;
-
-  constructor(props) {
-    super();
-    this.isControlled = props.open != null;
-    this.state = {
-      open: null,
-    };
-
-    if (!this.isControlled) {
-      // not controlled, use internal state
-      this.state.open = false;
+function Tooltip(props) {
+  const {
+    children,
+    classes,
+    disableFocusListener,
+    disableHoverListener,
+    disableTouchListener,
+    enterDelay,
+    enterTouchDelay,
+    id,
+    interactive,
+    leaveDelay,
+    leaveTouchDelay,
+    onClose,
+    onOpen,
+    open: openProp,
+    placement,
+    PopperProps,
+    theme,
+    title,
+    TransitionComponent,
+    TransitionProps,
+    ...other
+  } = props;
+
+  const [openState, setOpenState] = React.useState(false);
+  const [, forceUpdate] = React.useState(0);
+  const [childNode, setChildNode] = React.useState();
+  const ignoreNonTouchEvents = React.useRef(false);
+  const { current: isControlled } = React.useRef(props.open != null);
+  const defaultId = React.useRef();
+  const closeTimer = React.useRef();
+  const enterTimer = React.useRef();
+  const leaveTimer = React.useRef();
+  const touchTimer = React.useRef();
+  // can be removed once we drop support for non ref forwarding class components
+  const handleOwnRef = React.useCallback(ref => {
+    setChildNode(ReactDOM.findDOMNode(ref));
+  }, []);
+  const handleRef = useForkRef(children.ref, handleOwnRef);
+
+  React.useEffect(() => {
```

Comment:
This is dead code that won't be removed in production, it's also a pattern the future `warn()` method will not provide the component stack. I think that we should remove the effect. childNode.disabled can change without childNode changing. I'm not sure it deduplicate the logs correctly. Maybe we can start without the deduplicate then work something out later? https://github.com/bvaughn/react-window/blob/89287838dc7960e4838081786e0a4423f6607029/src/createListComponent.js#L110-L119.

Reply:
> This is dead code that won't be removed in production,

Can always wrap an `if (__DEV__)` around it if you think it's worth optimizing `a(function () {}, [b,c,d])`. Since it's a build time branch we can disable `rules-of-hooks`.

Otherwise I don't think we should remove warnings because they have false negatives i.e. don't detect all issues. If they at least detect issues sometimes they're useful. Usually it happens on the initial mount.

Reply:
I don't want to remove the warnings. I want to make them actionable. We still miss the component stack trace. I think that we should remove the usage of useEffect.